### ISSUE TITLE
TRN-4 — Safetensors checkpointing/resume and the DDP port

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1578]
**Depends on:** TRN-1 (#1575) · **Blocks:** GATE-3 (#1585), TRN-5 (#1639)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy training state is pickle end to end. `mace/tools/checkpoint.py` (235 lines) bundles `{model, optimizer, lr_scheduler}` state_dicts into a dict (`CheckpointBuilder.create_checkpoint`, `:29-34`) and `torch.save`s it as `<tag>_epoch-<N>[_swa].pt` (`:74-90`, `:162-174`), with the epoch encoded in the filename and the stage encoded as a `_swa` suffix parsed back by regex (`:100-122`); the final model is a **fully pickled `nn.Module`** (`torch.save(model_to_save, model_path)`, `mace/cli/run_train.py:1183`), which is why `mace/__init__.py:5` must set `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD=1` — the pickle security concern this ticket kills. Three structural defects to not reproduce: (1) the EMA shadow is never serialized — checkpoints are written *inside* `ema.average_parameters()` (`mace/tools/train.py:329-355`) so the file holds averaged weights and resume silently resets EMA; (2) checkpoints are backend-specific state_dicts, one reason five cross-backend converter CLIs exist (`mace/cli/convert_{e3nn_cueq,cueq_e3nn,e3nn_oeq,oeq_e3nn,e3nn_hybrid}.py`); (3) resume is a stack of bare `try/except Exception` blocks (`run_train.py:942-966`) that fall through swa → non-swa → "assume this is an LBFGS restart", so a genuinely corrupt checkpoint is indistinguishable from a stage mismatch. Legacy did gain one honest improvement worth keeping: a checkpoint whose optimizer param groups no longer match now restores the **weights** and logs why the optimizer/scheduler state was dropped (`checkpoint.py:41-51`) instead of dying — v1 reproduces that as an explicit, typed outcome rather than a warning. DDP lives in `mace/tools/distributed_tools.py` (66 lines: `slurm | torchrun | mpi | none` launchers; `nccl` for cuda, `ccl` for xpu, and a `gloo` CPU fallback so `--distributed` on CPU works at all) and `mace/tools/slurm_distributed.py` (51 lines: SLURM env → `RANK`/`WORLD_SIZE`/`MASTER_ADDR`). `run_train.py` wires it at `:108` (init), `:752-770` (per-head `DistributedSampler`), `:954-956` (DDP wrap, `device_ids` only for cuda/xpu — CPU/gloo requires `device_ids=None`), and again at `:1122` for test-set evaluation; the loop gates checkpointing on `rank == 0`, barriers each epoch (`train.py:264,:355`), and broadcasts the patience early-exit flag from rank 0 (`train.py:359-362`). This ticket ports checkpoint I/O and DDP onto TRN-1 (#1575)'s explicit training loop, in `mace_torch/train/checkpoint.py`, `mace_torch/train/ddp.py`, `mace_torch/train/slurm.py`, and `mace_torch/data/distributed_sampler.py`.

**Interface & constraints:**

1. **Checkpoint format — safetensors + JSON sidecar, zero pickle.** Tensors (model weights, optimizer/scheduler/EMA state) go in safetensors; all structure (param-group topology, epoch/stage counters, provenance) goes in the JSON sidecar. Never a pickled module, never `jit.save`, never a backend-specific `state_dict`: model weights are stored in the **canonical layout**, with the internal-weight ops' `to_canonical()`/`load_canonical()` applied **only** at this save/load boundary (the BKD-1 (#1560) contract) — so one checkpoint loads into any backend, the five cross-backend converters stay dead, and the `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD` hazard does not exist in v1. Checkpoints are device-agnostic; the legacy `--save_cpu` flag (`arg_parser.py:1085`, applied at `run_train.py:1181-1182`) is a recorded DROP, not a port.
2. **Resume is complete or it is an error.** A training checkpoint carries everything the loop needs to continue: canonical model weights (**raw**, not EMA-averaged), optimizer state, LR-scheduler state, the **EMA shadow as first-class serialized state** (fixing the legacy reset-on-resume defect), and epoch + stage counters in metadata (replacing the filename-parsing `_epoch-<N>` / `_swa` scheme; stage two is a recorded stage per TRN-2 (#1576)'s per-stage schedules). The resolved config is saved into the checkpoint metadata (the CFG-1 (#1574) mandatory-metadata rule). The one legitimate partial resume — optimizer param groups changed since the checkpoint was written (`checkpoint.py:41-51`) — is a **typed, reported outcome** (`ResumeResult.optimizer_state = "reinitialized"` with the reason), never a bare `except Exception`.
3. **Resume contract (pinned by the P0-5 E2E CLI contracts):** `--restart_latest` resumes from the newest checkpoint for the run tag; `--keep_checkpoints` and `--save_all_checkpoints` keep their legacy semantics (delete-old-by-default at `checkpoint.py:165-167`; best-so-far always written, all-epochs optional — `train.py:320-356`).
4. **DDP port.** Keep the four launcher modes `slurm | torchrun | mpi | none` (`--distributed` and `--launcher` are KEEP flags, `arg_parser.py:79-89`) and the per-device process-group backend: `nccl` (cuda), `ccl` (xpu), `gloo` (cpu). Keep the `device_ids` rule: it is valid only for single-device accelerator modules, so cuda passes `[local_rank]`, xpu passes `[xpu_device_index(local_rank)]` and CPU passes `None` — narrowing that to cuda alone silently gives XPU runs a CPU-style DDP (`run_train.py:980-990` records exactly that regression). `DistributedSampler` per dataloader with `set_epoch` each epoch (legacy `train.py:247`).
5. **Rank discipline:** checkpoints are written by **rank 0 only**, into a `run_dir` assumed to be on a shared filesystem; a barrier separates the write from any rank reading it back; control decisions (patience early-exit, stage switch) are made on rank 0 and broadcast, so ranks never diverge (legacy `train.py:327-328,:356-359`). This same rank-0-writes/all-ranks-load invariant is the substrate FT-2 (#1581)'s pseudolabel generation relies on — TRN-4 (#1578) provides it, FT-2 (#1581) consumes it.

**Task:**
1. `mace_torch/train/checkpoint.py`: checkpoint save/load/latest-discovery on safetensors + JSON sidecar; canonical weights via `to_canonical`/`load_canonical` at the boundary; epoch/stage/provenance and resolved config in the sidecar; the typed `ResumeResult`.
2. Wire it into TRN-1 (#1575)'s explicit loop: best-so-far + optional all-epochs checkpointing, `--restart_latest` resume restoring model, optimizer, scheduler, EMA shadow, and epoch/stage counters.
3. Serialize the EMA shadow alongside raw weights (never write EMA-averaged weights as "the" model state in a training checkpoint).
4. **Write the deliverable model artifact, which is a different thing from the resume checkpoint.** The resume checkpoint holds raw weights plus the EMA shadow, per constraint 2. The deliverable — canonical weights plus the JSON model sidecar, the pair every downstream consumer loads and the one `mace export` copies — is written at best-so-far and at end of run with the **EMA-averaged** weights whenever EMA is enabled, matching v0.3.x, where the deployed model is the EMA average by construction: the checkpoint is written under `ema.average_parameters()` (`mace/tools/train.py:329-355`), and the final artifact is that checkpoint loaded back and re-saved (`mace/cli/run_train.py:1137-1141,:1164-1183`). The averaging is applied once at that write and never mutates the resume checkpoint.
5. `mace_torch/train/ddp.py` + `mace_torch/train/slurm.py`: port `init_distributed` (slurm/torchrun/mpi/none) and the SLURM env resolver, keeping the gloo CPU branch and the cuda/xpu `device_ids` rule; `mace_torch/data/distributed_sampler.py` for the sampler wiring.
6. Rank-0 checkpoint gating, epoch barriers, and rank-0 broadcast of early-exit/stage decisions in the loop.
7. CPU-DDP smoke under `--engine v1`: 2-process gloo training (pattern of develop's `tests/workflows/test_distributed.py`: spawn both ranks directly with a static rendezvous on `127.0.0.1:<free port>` and `GLOO_SOCKET_IFNAME=lo0` on darwin / `lo` elsewhere, so the test never depends on host DNS), asserting both ranks exit 0 and exactly one checkpoint set exists; extend it with a kill-and-resume leg.

**Out of scope:** multi-dataloader balancing and metrics/logging (TRN-3 (#1577)); the L-BFGS DDP protocol, which is its own broadcast dance (TRN-5 (#1639)); pseudolabel generation itself (FT-2 (#1581) — this ticket only guarantees the shared-run_dir + barrier substrate); the foundation-model one-shot converter and the neutral-format export of published artifacts (FM-00 (#1565)/FM-1 (#1586)); GPU (nccl) DDP testing — needs a multi-GPU runner, stays out of the smoke.

**Acceptance criteria:**
- [ ] The v1 checkpoint path contains no pickle: files are safetensors + JSON only; no `torch.save`/`torch.load` of objects, no `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD` anywhere in v1.
- [ ] Save→load round trip restores model, optimizer, LR-scheduler, EMA shadow, and epoch/stage counters exactly (state equality at fp64), and the reloaded model reproduces the pre-save outputs.
- [ ] A run interrupted mid-training and resumed with `--restart_latest` continues from the recorded epoch with EMA evolving from its restored shadow (not reset) — asserted by comparing EMA state across the resume boundary.
- [ ] A checkpoint whose optimizer param groups no longer match loads its weights and reports `optimizer_state = "reinitialized"` with a reason; a corrupt checkpoint raises instead of falling through to the next candidate.
- [ ] `--keep_checkpoints` / `--save_all_checkpoints` behave as in legacy; `--save_cpu` is absent from the v1 surface.
- [ ] Checkpoint weights are canonical: a checkpoint written under one backend loads under another registered backend via the registry (BKD-1 (#1560)) with no conversion step.
- [ ] CPU-DDP smoke passes: 2 ranks on gloo train through the stage-two switch, both ranks return 0, only rank 0 wrote checkpoints, and a DDP run resumes cleanly from them.
- [ ] The stage counter recorded in metadata replaces filename parsing — no `_epoch-`/`_swa` regexes in v1.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/train/test_checkpoint.py -v      # round trip, resume, EMA-across-resume, no-pickle
python -m pytest tests/workflows/test_distributed.py -v                     # 2-process gloo smoke, re-run under --engine v1
```
**Review focus:** resume completeness — nothing silently reinitialized (optimizer param groups, scheduler counters, EMA shadow, stage state) — and the rank discipline: rank-0-only writes with a barrier before any read-back, broadcast control decisions, no partial checkpoints on interrupt.

---
